src/lib/ndpi_main.c

```diff
@@ -8546,13 +8546,27 @@ static ndpi_protocol ndpi_internal_detection_process_packet(struct ndpi_detectio
   /* Zoom cache */
   if((ret.app_protocol == NDPI_PROTOCOL_ZOOM) && (flow->l4_proto == IPPROTO_TCP))
     ndpi_add_connection_as_zoom(ndpi_str, flow);
-  
+
   if(ndpi_str->cfg.fully_encrypted_heuristic &&
      ret.app_protocol == NDPI_PROTOCOL_UNKNOWN && /* Only for unknown traffic */
      flow->packet_counter == 1 && packet->payload_packet_len > 0) {
    flow->first_pkt_fully_encrypted = fully_enc_heuristic(ndpi_str, flow);
   }
 
+  if(flow->first_pkt_fully_encrypted == 0 &&
+     ret.app_protocol == NDPI_PROTOCOL_UNKNOWN &&
+     NDPI_ENTROPY_ENCRYPTED_OR_RANDOM(flow->entropy) == 0 &&
+     flow->packet_counter < 3)
+  {
+    flow->entropy = ndpi_entropy(packet->payload, packet->payload_packet_len);
+    if(NDPI_ENTROPY_ENCRYPTED_OR_RANDOM(flow->entropy) != 0) {
+      char str[32];
+
+      snprintf(str, sizeof(str), "Entropy %.2f", flow->entropy);
+      ndpi_set_risk(flow, NDPI_SUSPICIOUS_ENTROPY, str);
+    }
+  }
+
```

Comment:
We end up having an entropy value even for classified traffic, as reported in the unit tests diff (trivial example: entropy calculated in the first packet, but flow classified only on the second pkt)

Is that what we want?

BTW: I am right saying that, if we have that value, it is always calculated on the first packet of the flow?

Reply:
Can you point me the diff test, because all I've found were Matches by IP/Port. I guess it's fine to calculate the entropy for all !DPI confidence flows.

Reply:
tests/cfgs/default/result/dnscrypt-v1-and-resolver-pings.pcap.out

I agree; no issues when the classification is !DPI